### PR TITLE
Adding a time series utility.

### PR DESCRIPTION
**Pull request by boks1971**
Also adding a `SWEEPER` node type.

Will add more comments in line.

Adding a time series utility module for keeping stats. Planning to use it in relay stats. I think it can be more generally usable. Could not find a good existing library. As it is simple enough and also have some unique requirements, just making one. But, if you know of any existing good ones, please let me know and I can change this and make this a wrapper.

### CODE REVIEW

File: utils/timeseries.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+				}
+				break
+			}
+		}
+	}
+
+	t.prune()
+}
+
+func (t *TimeSeries[T]) prune() {
```
Comment (boks1971) [inline utils/timeseries.go:214]: currently doing `prune` on both read and write.

File: utils/timeseries.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	thresh := time.Now().Add(-t.params.Window)
+	for idx, s := range t.samples {
+		if s.At.After(thresh) {
+			t.samples = t.samples[idx:]
+			break
+		}
+	}
+}
+
+// TODO - a bunch of stats
```
Comment (boks1971) [inline utils/timeseries.go:224]: currently supporting only `min, max, sum`. We can add more over time as the need arises.

File: livekit_internal.proto [hunk 3]
```diff
@@ -16,6 +16,7 @@ enum NodeType {
   CONTROLLER = 1;
   MEDIA = 2;
   TURN = 4;
+  SWEEPER = 5;
```
Comment (boks1971) [inline livekit_internal.proto:19]: Adding a node type which we can use in cloud.

**Comment by paulwe** [comment]
i think we might be able to use gonum for this

File: utils/timeseries.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+func (t *TimeSeries[T]) addSampleAt(val T, at time.Time) {
+	// insert in time order
+	switch {
+	case len(t.samples) == 0 || at.After(t.samples[len(t.samples)-1].At): // empty or at end
+		t.samples = append(t.samples, TimeSeriesSample[T]{
+			Value: val,
+			At:    at,
+		})
+
+	case at.Before(t.samples[0].At): // at start
```
Comment (cnderrauber) [inline utils/timeseries.go:187]: If there are many insert operation at start or middle to the samples, list might be better than slice.

**Review by cnderrauber — APPROVED** [review]

File: utils/timeseries.go (hunk 4 above)
Comment (boks1971) [reply utils/timeseries.go:187]: Good point @cnderrauber . Will change to a list.

**Comment by boks1971** [comment]
> i think we might be able to use gonum for this

I looked at that @paulwe , but could not find a time series in there. Will check for more options. We can always switch out later too.

**Pull request merged**